### PR TITLE
[SPARK-33503][SQL] Refactor SortOrder class to allow multiple childrens

### PR DESCRIPTION
**Pull request by prakharjain09**
### What changes were proposed in this pull request?
This is a followup of #30302 . As part of this PR, sameOrderExpressions set is made part of children of SortOrder node - so that they don't need any special handling as done in #30302 .

### Why are the changes needed?
sameOrderExpressions should get same treatment as child. So making them part of children helps in transforming them easily.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing UTs

### CODE REVIEW

**Comment by prakharjain09** [comment]
@cloud-fan @maropu I have made the changes as suggested in https://github.com/apache/spark/pull/30302#discussion_r526104333 .

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by maropu** [comment]
@prakharjain09 Thanks for working on this. btw, could you assign a new jira ID to this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala [hunk 1]
```diff
@@ -64,7 +64,9 @@ case class SortOrder(
     direction: SortDirection,
     nullOrdering: NullOrdering,
     sameOrderExpressions: Set[Expression])
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala:66]: We need to use `Set` for this variable?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala [hunk 2]
```diff
@@ -64,7 +64,9 @@ case class SortOrder(
...
-  extends UnaryExpression with Unevaluable {
+  extends Expression with Unevaluable {
+
+  override def children: Seq[Expression] = child +: sameOrderExpressions.toSeq
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala:70]: nit:
```
  def satisfies(required: SortOrder): Boolean = {
    (sameOrderExpressions + child).exists(required.child.semanticEquals) &&
      direction == required.direction && nullOrdering == required.nullOrdering
  }
```
=>
```
  def satisfies(required: SortOrder): Boolean = {
    children.exists(required.child.semanticEquals) &&
      direction == required.direction && nullOrdering == required.nullOrdering
  }
```
?

Comment (prakharjain09) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala:70]: done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala (hunk 1 above)
Comment (prakharjain09) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala:66]: The set property of this variable is used at couple of places in SortMergeJoin ([here](https://github.com/apache/spark/blob/3fdfce3120f307147244e5eaf46d61419a723d50/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala#L122), [here](https://github.com/apache/spark/blob/v3.0.0/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala#L149)).

We can use ExpressionSet maybe at those 2 places and make it a Seq in constructor here. any thoughts?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala:66]: > We can use ExpressionSet maybe at those 2 places and make it a Seq in constructor here. any thoughts?

Yea, I think it is okay to just deduplicate it before storing it in the class as you said, and then make it `Seq`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 3]
```diff
@@ -68,9 +68,9 @@ case class SortMergeJoinExec(
       val leftKeyOrdering = getKeyOrdering(leftKeys, left.outputOrdering)
       val rightKeyOrdering = getKeyOrdering(rightKeys, right.outputOrdering)
       leftKeyOrdering.zip(rightKeyOrdering).map { case (lKey, rKey) =>
-        // Also add the right key and its `sameOrderExpressions`
-        SortOrder(lKey.child, Ascending, lKey.sameOrderExpressions + rKey.child ++ rKey
-          .sameOrderExpressions)
+        // Also add expressions from right side sort order
+        val sameOrderExpressions = ExpressionSet(lKey.children ++ rKey.children) - lKey.child
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:72]: why not `ExpressionSet(lKey.sameOrderExpressions ++ rKey.children)`?

Comment (prakharjain09) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:72]: I was thinking if "lKey.child" can be part of rKey.children. But that shouldn't be the case as left and right side of SMJ will always return different expressions ids (even when they work on same tables).

making suggested change here.

**Review by maropu — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 4]
```diff
@@ -1090,6 +1090,34 @@ class PlannerSuite extends SharedSparkSession with AdaptiveSparkPlanHelper {
     }
   }
 
+  test("sort order doesn't have repeated expressions") {
+    withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1") {
+      withTempView("t1") {
+        withTempView("t2") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:1096]: nit: `withTempView("t1", "t2")`

**Review by cloud-fan — APPROVED** [review]

**Comment by maropu** [comment]
GA test passed, so I merged to master. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 4 above)
Comment (prakharjain09) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: done.

**Pull request closed without merging**